src/msg/MessageFactory.h

```diff
@@ -0,0 +1,72 @@
+//
+// Created by skyinno on 2/14/16.
+//
+
+#ifndef CEPH_MESSAGEFACTORY_H
+#define CEPH_MESSAGEFACTORY_H
+
+#include <map>
+#include <functional>
+#include <memory>
+#include "Message.h"
+
+struct MessageFactory
+{
+    template<typename T>
+    struct register_t
+    {
+        register_t(int key)
+        {
+            MessageFactory::get().map_.emplace(key, &register_t<T>::create);
+        }
+
+        template<typename... Args>
+        register_t(int key, Args... args)
+        {
+            MessageFactory::get().map_.emplace(key, [&] { return new T(args...); });
+        }
+        inline static Message* create() { return new T; }
+    };
+
+    inline Message* produce(int key)
+    {
+        if (map_.find(key) == map_.end())
```

Comment:
if you store the iterator returned by `find()`, you can avoid doing another binary search below with `map_[key]`